### ISSUE TITLE
Tasker Setup Troubleshooting

### ISSUE THREAD

**Issue by jun3280net** [adriankumpf/teslamate#102]
As noted in the guide, @adriankumpf setup Teslamate with dynamic polling minimize the frequency your vehicle is polled when it is sleeping/inactive.  

Tasker can automatically trigger Teslamate to resume polling the vehicle upon connecting to your Tesla's bluetooth.

With guidance from @IMgoRt :
"In tasker open a new task, add a javascriptlet and put the following:
xmlhttp = new XMLHttpRequest(); xmlhttp.open('PUT', 'https://192.168.1.1:4000/api/car/1/logging/resume');
where 192,168.1.1 is the IP address (I have mine over a VPN), 4000 is the port and 1 is the car number"

![image](https://user-images.githubusercontent.com/30555695/63231872-0fe45200-c1df-11e9-829f-c9a3b8f73a74.png)

Unfortunately, this did not work for me - perhaps related to my NGINX setup.  Perhaps I don't have Tasker configured properly yet.  Maybe others can comment on what needs to be done to make this work.

I can however confirm this works without Secure HTTP.

Create a new Task : "Teslamate Resume"
Click the + sign to add a new **Action>Net>HTTP Request**
Set the Method to **PUT**
Set the URL to http://192.168.86.22:4000/api/car/1/logging/resume

![image](https://user-images.githubusercontent.com/30555695/63232016-c563d500-c1e0-11e9-86b4-1e24661c1593.png)

You can then test it within Tasker by clicking the play button and checking against the Teslamate Server log.  

![image](https://user-images.githubusercontent.com/30555695/63231984-7322b400-c1e0-11e9-8eb4-cc183d03f45c.png)

Trigger the Task by connecting to your Model S Bluetooth 
Select **PROFILES**
Click the + on the bottom right and name it : BT Connected
Choose **STATE>Net>BT Connected**
     Name : click the magnifying glass to choose your Tesla
     Address : click the magnifying glass to choose your Tesla
Select the Task Teslamate Resume

![image](https://user-images.githubusercontent.com/30555695/63232199-ce55a600-c1e2-11e9-8979-c8394302165a.png)

![image](https://user-images.githubusercontent.com/30555695/63232175-94849f80-c1e2-11e9-97b6-a2c9e5f554c4.png)

Setup a similar process to SUSPEND as required.

Hopefully others will find this useful.  In the meantime, I will try and figure out how to make this work outside of my LAN.

**Comment by Simple-Complexity** [comment]
To me it makes sense that it would not work over https, but it would over http. Teslamate, by default, listens to http requests (as far as I'm aware) and unless you have something else listening for https requests and terminating SSL it will not work.

Given that you are using the same port for both http and https it seems to me that it unlikely you have another NGINX server setup to accept https requests as that would cause port conflicts.

Maybe I'm missing something?

**Comment by IMgoRt** [comment]
Thanks @jun3280net for this.
I can confirm that using JavaScriptlet works over HTTPS whereas using HTTP Request only works over HTTP in my environment too (OpenVPN access)

I've no idea why.

**Comment by DrMichael** [comment]
Hmmm, I cannot get it working. I get

teslamate_1  | 15:37:51.802 [info] GET /api/car/1/logging/resume
teslamate_1  | 15:37:51.802 [info] Sent 404 in 247µs
teslamate_1  | 15:37:51.802 [info] Converted error Phoenix.Router.NoRouteError to 404 response

when I try to access http://<myhost>:4000/api/car/1/logging/resume

**Comment by virtualm2000** [comment]
> Hmmm, I cannot get it working. I get
> 
> teslamate_1 | 15:37:51.802 [info] GET /api/car/1/logging/resume
> teslamate_1 | 15:37:51.802 [info] Sent 404 in 247µs
> teslamate_1 | 15:37:51.802 [info] Converted error Phoenix.Router.NoRouteError to 404 response
> 
> when I try to access http://:4000/api/car/1/logging/resume

I'm not familiar with Tasker but you need a PUT request not GET. Make sure you select PUT Method.

**Comment by bahree** [comment]
In addition is your IP correct or you have just removed it for posting here?

**Comment by DrMichael** [comment]
Ah, ok, that gives me

teslamate_1  | 16:05:07.309 [info] PUT /api/car/1/logging/resume
teslamate_1  | 16:05:07.313 [info] Sent 204 in 3ms

What probably means, that it is ok...

**Comment by DrMichael** [comment]
> In addition is your IP correct or you have just removed it for posting here?

I replaced it by <myhost>, but that apparently got lost...

**Comment by jun3280net** [comment]
Ok!  I finally got something that should works well enough.  Anybody more technical inclined in the space of network security - feel free to chime in, as I'm new to all of this.  I was largely unsuccessful playing get NGINX to do what I needed it to do, so I started to simply look into setting up a VPN.  

I spun up another instance of Ubuntu on VMWARE (I'm sure I could use the same image I have for TESLAMATE - but didn't want to risk corrupting it).  That said, with the new instance of VMWARE - simply follow the guide here which takes no longer than 10 minutes to get OPENVPN server setup to allow you to connect to your LAN from the outside.

[https://www.cyberciti.biz/faq/howto-setup-openvpn-server-on-ubuntu-linux-14-04-or-16-04-lts/]

Once you have the server setup, be sure to port forward your servers IP on port 1194 on your router.

Using your phone, disconnect from your LAN (eg shut off wifi to use LTE).  

Install OPENVPN client on your phone and import the file.ovpn file your server created (I emailed this file to myself).  Add the connection, and it should load up. 

Once it is connected, you should be able to type your TESLAMATE's LAN IP (eg 192.168.86.22:4000) into your Chrome browser.  Voila - you are connected to your server from the outside.

Use the TASKER setup I provided above to get your Tesla to trigger the Teslamate server each time your bluetooth connects to the car.

I hope this helps - and feel free to comment on how to improve this setup.

**Comment by adriankumpf** [comment]
Thanks @jun3280net! I added a link to your comment to this new wiki page: [Community Guides](https://github.com/adriankumpf/teslamate/wiki/Community-Guides) :)

**Comment by jun3280net** [comment]
> Thanks @jun3280net! I added a link to your comment to this new wiki page: [Community Guides](https://github.com/adriankumpf/teslamate/wiki/Community-Guides) :)

Thanks @adriankumpf keep up the great work!  Look forward to the future features down the pipe.

**Comment by spacecosmos** [comment]
Any luck with getting the Tasker to work when using HTTPS connection @jun3280net ? I'm using the Docker installation of TeslaMate and would love to wake up the connection when it senses and connects my phone via bluetooth.

Here's my guide on the Docker setup at [DigitalOcean](https://m.do.co/c/98c628ea4a8e):
[https://satheesh.net/2019/09/28/teslamate-digitalocean-docker-step-by-step-installation-guide/](url)

When using HTTPS the port number isn't specified after the FQDN. Is there a way to access the command `/api/car/1/logging/resume` differently perhaps? Web-server? Maybe @adriankumpf have some tips?

**Comment by adriankumpf** [comment]
I'm not familiar with Tasker but in theory you only need to set the Authorization header with your username and password (from `.htpasswd`):

```javascript
xmlttp.setRequestHeader("Authorization", "Basic " + btoa("yourusername:yourpassword"));
```
The port doesn't need to be specified. Just use your domain name:

```
https://teslamate.example.com/api/car/1/logging/resume
```

**Comment by spacecosmos** [comment]
> 
> The port doesn't need to be specified:
> 
> ```
> https://teslamate.example.com/api/car/1/logging/resume
> ```
@adriankumpf 
I tried putting _**just**_ the url _**with**_ `/api/car/1/logging/resume` in the browser address without success. Maybe I'm missing something? Is the "car/1" correct or is the value supposed to be something else? Since I'm already authenticated I would think it would work without specifying any additional credentials as long as I'm just opening another tab in the same browser session.

**Example:**
`https://mydomain.com/api/car/1/logging/resume`

**This gives me this error in the browser windows:**
Not Found

**Comment by adriankumpf** [comment]
It cannot simply be opened in the browser because it must be a PUT request.

**Comment by spacecosmos** [comment]
> It cannot simply be opened in the browser because it must be a [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT) request.

Ok - then my developer skills are a bust at this point. :)

PS! Maybe you can implement a link on the TeslaMate Dashboard? 💯

**Comment by jun3280net** [comment]
> Any luck with getting the Tasker to work when using HTTPS connection @jun3280net ? I'm using the Docker installation of TeslaMate and would love to wake up the connection when it senses and connects my phone via bluetooth.
> 
> Here's my guide on the Docker setup at [DigitalOcean](https://m.do.co/c/98c628ea4a8e):
> [https://satheesh.net/2019/09/28/teslamate-digitalocean-docker-step-by-step-installation-guide/](url)
> 
> When using HTTPS the port number isn't specified after the FQDN. Is there a way to access the command `/api/car/1/logging/resume` differently perhaps? Firewall issue maybe? Web-server? Maybe @adriankumpf have some tips?

hi @spacecosmos  - unfortunately, I was not able to get HTTPS working.  I couldn't get NGINX working given my very limited understanding with networking and linux.  As I was able to get it working using HTTP....I ended up just setting up a VPN.  That worked best for me - because whenever I'm connected to the VPN, I can see my local area network.  

On a separate note - for HTTPS - did you setup a DNS name?  I think if you have a DNS name wouldn't you be able to bypass the port number that way?

@adriankumpf  used to have a link to wake up /sleep the car on Teslamate (Port 4000), but I don't see it anymore.

**Comment by adriankumpf** [comment]
> @adriankumpf used to have a link to wake up /sleep the car on Teslamate (Port 4000), but I don't see it anymore.

You mean the button to suspend / resume logging, right? It's still there. But it's not visible all the time, only when it makes sense e.g. if the car is online and locked or if it is asleep.

**Comment by jun3280net** [comment]
> > @adriankumpf used to have a link to wake up /sleep the car on Teslamate (Port 4000), but I don't see it anymore.
> 
> You mean the button to suspend / resume logging, right? It's still there. But it's not always visible, only when it makes sense, like when the car is online and locked, or when it's attempting to sleep.

Weird - honestly I haven't seen it in a very long time (since like V 1.7 I think).

**Comment by spacecosmos** [comment]
> 
> On a separate note - for HTTPS - did you setup a DNS name? I think if you have a DNS name wouldn't you be able to bypass the port number that way?
> 
@jun3280net 
Yes, I do have a FQDN (fully qualified domain name). And everything is working pretty much perfectly! The only thing I want now is to be able to resume Teslamate when my car connects to he bluetooth.

But as @adriankumpf mentions, the PUT command is probably needed and I'm totally lost on how to get that to work on Tasker (or a website for that matter).

If anyone else out there has any pointers I would be very grateful!

**Comment by jun3280net** [comment]
> > On a separate note - for HTTPS - did you setup a DNS name? I think if you have a DNS name wouldn't you be able to bypass the port number that way?
> 
> @jun3280net
> Yes, I do have a FQDN (fully qualified domain name). And everything is working pretty much perfectly! The only thing I want now is to be able to resume Teslamate when my car connects to he bluetooth.
> 
> But as @adriankumpf mentions, the PUT command is probably needed and I'm totally lost on how to get that to work on Tasker (or a website for that matter).
> 
> If anyone else out there has any pointers I would be very grateful!

@spacecosmos  - dumb question, but you copied the very first picture I posted - which was based on what @IMgoRt  mentioned?  

"In tasker open a new task, add a **javascriptlet** and put the following:
xmlhttp = new XMLHttpRequest(); xmlhttp.open('PUT', 'https://192.168.1.1:4000/api/car/1/logging/resume');

That was using HTTPS for him using the PUT command.  @IMgoRt  mentioned it took him a while to debug - but this worked for him.

**Comment by spacecosmos** [comment]
> Thanks @jun3280net for this.
> I can confirm that using JavaScriptlet works over HTTPS whereas using HTTP Request only works over HTTP in my environment too (OpenVPN access)
> 
> I've no idea why.

@jun3280net 
Are you using a local server or a public reachable server with dns name? I just cannot get it to work with a public dns name. Why need to specify the port number when running HTTPS? I'm totally confused...

(PS! My coding skills are seriously bad.)

**Comment by spacecosmos** [comment]
I actually found this from the original [Tesla Logger](http://teslalogger.de/handbuch.php) manual in german.

I put it though Google Translate and did some minor tweaks and got this: [Wake up with Tasker.pdf](https://github.com/adriankumpf/teslamate/files/3680065/Wake.up.with.Tasker.pdf)

> **Wake up with Tasker**
> «Tasker» is a paid app for Android smartphones that are in the Google Play Store. Allows activities based on events on the phone to start automatically.
> 
> Prerequisite: The phone is via Bluetooth with the vehicle as a hands-free device coupled and configured not only via the Tesla app as a car key.
> 
> With Tasker we catch the moment when the phone connects via Bluetooth to the car and then wake up Tesla Logger, as apparently now the vehicle is to be moved. But then it has to happen within the next 15 minutes.
> 
> **This is how «Tasker» is configured for this use case:**
> 1. In the admin area of the Tesla Logger, display the log file and search for the lines in where the following lines can be found:
> 
> Server Port: https://teslalogger.de
> Path: wakeup.php
> Attributes: t = 1234abcd
> (the value behind "t =" is individual for the vehicle)
> 
> 2. Start Tasker on the phone and add a new "Profile", select "Status", «Network», then «Bluetooth connected». In the Name field, select the name of the Tesla. This can only be selected there if the phone is previously paired via Bluetooth has been (see above).
> 
> 3. Select «New task» for the profile and select a name, ie «Teslalogger»
> 
> 4. Select «+», «Network», then «HTTP Request»
> <img width="290" alt="Screenshot 2019-10-02 at 09 57 55" src="https://user-images.githubusercontent.com/55928670/66028769-0ba99500-e4fe-11e9-8e27-6866a093c5ee.png">
> 
> 5. Under "URL" it must be "https://teslalogger.de/wakeup.php" (case sensitive is important!). Under "Query Parameters" must be what is in the logfile when restarting from the Tesla logger, so for example "t = 0" (as shown in the picture). All other statements remain as they are.
> 
> 6. Tasker must always be running, as indicated by a small icon in the bar at the top left.




He uses a `wakeup.php` file on the webserver to get it to work. Maybe this is a solution @jun3280net?

**Comment by MertonKnight** [comment]
Hi @spacecosmos 

The wakeup.php file isn't going to help since it's doesn't exist in Teslamate.

I read your documentation on how you setup Teslamate in DigitalOcean. If you followed the advanced steps for setting up the .htpasswd file you need to install a Tasker plugin so you can authenticate using basic access authentication. I couldn't get authentication to work in Tasker without this plugin.

https://play.google.com/store/apps/details?id=com.freehaha.restask&hl=en_US

Let me know if you need more detailed instructions. I basically have the same setup as you except it's in AWS.

**Comment by spacecosmos** [comment]
> Hi @spacecosmos
> 
> The wakeup.php file isn't going to help since it's doesn't exist in Teslamate.
> 
> I read your documentation on how you setup Teslamate in DigitalOcean. If you followed the advanced steps for setting up the .htpasswd file you need to install a Tasker plugin so you can authenticate using basic access authentication. I couldn't get authentication to work in Tasker without this plugin.
> 
> https://play.google.com/store/apps/details?id=com.freehaha.restask&hl=en_US
> 
> Let me know if you need more detailed instructions. I basically have the same setup as you except it's in AWS.

@MertonKnight, thanks for this and great to have someone who's in pretty much the same situation as I. But my skills with Tasker (and Docker) isn't all that good I'm afraid. So some exact guidelines from you would be helpful. Screenshots and examples would be really helpful.

Thanks again for reaching out!

**Comment by ngardiner** [comment]
I'm working on step by step instructions with screenshots, give me a few days and I'll have it ready.

**Comment by spacecosmos** [comment]
Awesome @ngardiner!!

**Comment by MertonKnight** [comment]
@spacecosmos 

Here are instructions for creating Tasker tasks to resume and suspend logging when Teslamate is behind a reverse-proxy with HTTPS and basic access authentication. Give these a try. Let me know if you run into any issues. Once you get this working you can create a profile that will automate running the tasks.

1. Install this plug-in for Tasker
https://play.google.com/store/apps/details?id=com.freehaha.restask&hl=en_US

2. Open Tasker. 

3. Do the following as shown in the picture below:
-Select the **TASKS** tab. 
-Add a task by pressing the plus sign in the bottom right. 
-Name the task **TeslamateResume**.
-Press the checkmark.
![Screenshot_20191006-181248-1small](https://user-images.githubusercontent.com/56217254/66277790-59612d00-e871-11e9-81fb-1be94e7be8f4.png)

4. On the Select Action dialog do the following as shown below:
-Type **rest** in the filter field
-Select **RESTask**
![Screenshot_20191006-181312-2small](https://user-images.githubusercontent.com/56217254/66277892-4e5acc80-e872-11e9-9be2-b415d4850ae3.png)


5. On the **Action Edit** screen press the configuration edit button as shown below:
-Optional: Set timeout to 30 seconds before editing.
![Screenshot_20191006-181321-3small](https://user-images.githubusercontent.com/56217254/66277979-0be5bf80-e873-11e9-8e1e-0d68073f5392.png)


6. On the **RESTask** screen do the following:
-Change **Request Type** to **PUT**.
-Change **Host** to **https://YOUR_FQDN/api/car/1/logging/resume** . Replace YOUR_FQDN with what FQDN_TM is set to in the .env file.
-Change **Basic Auth** to the credentials used for the .htpasswd file.
![Screenshot_20191006-181635-4small](https://user-images.githubusercontent.com/56217254/66277987-1e5ff900-e873-11e9-8155-7a3791b3e2d1.png)

7. Press to play button to test. You should get a 204 response if it's successful and also something similar to below will appear in the Docker logs. If this does not appear double check your settings in step 6.
![Capture](https://user-images.githubusercontent.com/56217254/66278001-38014080-e873-11e9-974e-cff002513f0d.PNG)

8. Press the save button. Go back to the Tasker home screen

9. Repeat steps 3 to 8 and create a suspend task make sure to do the following:
-On Step 3 use the name **TeslamateSuspend**
-On Step 6 use **https://YOUR_FQDN/api/car/1/logging/suspend** for the host

Once you get the above working you can create a profile that will automate running the tasks based on connecting to the car's Bluetooth. Let me know if you need help with that. Instructions are kind of in the first post.

**Comment by spacecosmos** [comment]
@MertonKnight - Thank you for taking the time to do this. It worked instantly! Extremely excited and grateful for this and I'm sure others within this community are too. 

(I hope you don't mind me publishing this alongside **[my guide](https://satheesh.net/2019/09/28/teslamate-digitalocean-docker-step-by-step-installation-guide/)** to get TeslaMate up and running on DigitalOcean?)

**Comment by MertonKnight** [comment]
@spacecosmos - That's great to hear. Glad you got it working. Yes, you can publish it in your guide. Don't need to give me credit. Also, I'm planning to update my steps. I figured out how to create the task without without a plug-in. I'll post that within a day.

Having the suspend task is good to have. It will help your Tesla go to sleep faster and reduce vampire drain.

One small cosmetic suggestion for your profile setup. Instead of having two separate profiles you can combine them into one by creating an **Exit Task**. If you press hold on TM Resume you'll get some options.
![Screenshot_20191007-005822-small](https://user-images.githubusercontent.com/56217254/66286360-2170de80-e89f-11e9-973e-64aa8f8097f1.png)

Tap **Add Exit Task** so can add a task for when the Bluetooth connection to your car disconnects. It'll look like this after you add it.
![Screenshot_20191007-010415-small](https://user-images.githubusercontent.com/56217254/66286413-63018980-e89f-11e9-9935-7a3d343cbf91.png)

This does the same thing as what you already have but gets it all in one profile.

**Comment by spacecosmos** [comment]
@MertonKnight - Again, worked like a charm! Thanks a lot for all the help! I also think it's best to combine them actually. Because it seems that the **Suspend** task runs all the time on my phone when it detects that I'm _not_ connected to the cars Bluetooth. But now, it should only work when the **Resume** task is "finished running" and disconnected I guess? 
![SmartSelect_20191007-101431_Tasker](https://user-images.githubusercontent.com/55928670/66295519-d4ffc000-e8eb-11e9-8892-66d21a7d42c6.jpg)

Next I want to try tapping an NFC-tag placed somewhere in the dash. (But that's for future testing...)
**:)**

**Comment by spacecosmos** [comment]
Updated the guide to include the Tasker App setup as well. Thank you all for helping out! 💯
**https://satheesh.net/2019/09/28/teslamate-digitalocean-docker-step-by-step-installation-guide/**

**Comment by adriankumpf** [comment]
The new [documentation](https://github.com/adriankumpf/teslamate/tree/master/docs) hopefully answers many of the questions that have been asked in this thread. If you have any further questions that aren't answered in the docs, don't hesitate to open a new issue.
